backend/src/main/java/gov/cdc/usds/simplereport/service/UploadService.java

```diff
@@ -1,92 +1,111 @@
+            throw new IllegalGraphqlArgumentException("Empty or invalid CSV submitted");
         }
 
+        int rowNumber = 0;
         while (valueIterator.hasNext()) {
-            final Map<String, String> row = valueIterator.next();
-
-            final LocalDate patientDOB = LocalDate.parse(row.get("DOB"), DATE_FORMATTER);
-            final String phone = normalizePhoneNumber(row.get("PhoneNumber"));
-            _ps.addPatient(
-                    row.get(FACILITY_ID).equals("") ? null : UUID.fromString(row.get(FACILITY_ID)),
-                    null,
-                    row.get("FirstName"),
-                    row.get("MiddleName"),
-                    row.get("LastName"),
-                    row.get("Suffix"),
-                    patientDOB,
-                    row.get("Street"),
-                    row.get("Street2"),
-                    row.get("City"),
-                    row.get("State"),
-                    row.get("ZipCode"),
-                    phone,
-                    row.get("Role").toUpperCase(),
-                    row.get("Email"),
-                    row.get("County"),
-                    row.get("Race").toLowerCase(),
-                    ethnicityMap.get(row.get("Ethnicity")),
-                    row.get("Gender").toLowerCase(),
-                    yesNoMap.get(row.get("residentCongregateSetting")),
-                    yesNoMap.get(row.get("employedInHealthcare")));
+            final Map<String, String> row = getNextRow(valueIterator);
+            rowNumber++;
+            try {
+                _ps.addPatient(
+                    parseUUID(row.get(FACILITY_ID)),
+                    null, // lookupID. this field is deprecated
+                    parseString(row.get("FirstName")),
+                    parseString(row.get("MiddleName")),
+                    parseString(row.get("LastName")),
+                    parseString(row.get("Suffix")),
+                    parseUserShortDate(row.get("DOB")),
+                    parseString(row.get("Street")),
+                    parseString(row.get("Street2")),
+                    parseString(row.get("City")),
+                    parseState(row.get("State")),
+                    parseString(row.get("ZipCode")),
+                    parsePhoneNumber(row.get("PhoneNumber")),
+                    parsePersonRole(row.get("Role")),
+                    parseEmail(row.get("Email")),
+                    parseString(row.get("County")),
+                    parseRace(row.get("Race")),
+                    parseEthnicity(row.get("Ethnicity")),
+                    parseGender(row.get("Gender")),
+                    parseYesNo(row.get("residentCongregateSetting")),
+                    parseYesNo(row.get("employedInHealthcare"))
+                );
+            } catch (IllegalGraphqlArgumentException e) {
+                throw new IllegalGraphqlArgumentException("Error on row "+ rowNumber+ "; " + e.getMessage());
```

Comment:
Do we have any size constraints on the CSV? Ideally we should be validating the whole CSV before sending any records into the PersonService. The downside with that approach would be the need to load the full CSV into memory before persisting any records.

One potential outcome with validating and persisting in the same iteration over the document is that an error gets encountered after one or more rows are successfully parsed, so if the user corrected the error in place and reattempted the upload, we would end up with duplicate records (or uniqueness errors if there's a constraint at the DB level). Spring may have some transactional magic going on that is already addressing this, though.

Reply:
Open to suggestions on how to limit upload size. Defiantly want some sort of limit before we expose it to the non admin users.

Right now there is some transaction magic going on. Will add a test to `UploadServiceTest` to illustrate this

Reply:
I think the App Services layer probably can be given a constraint for POST body size. 

* in Spring itself, multipart uploads can be limited using properties: https://stackoverflow.com/questions/34177873/max-limit-of-multipartfile-in-spring-boot/44332602#44332602
* I think more to the point, the standard tomcat connector settings (which we can tweak through the application props) allow setting maximum POST body size, which appears to default to 2MB: https://tomcat.apache.org/tomcat-8.0-doc/config/http.html#Common_Attributes

Reply:
I have a separate comment related to this, regarding field size. If we know we don't have essay-length fields (and I'm pretty sure we don't) we should prevent strings from being crazy long.